torch/csrc/distributed/c10d/FlightRecorder.cpp

```diff
@@ -334,208 +688,35 @@ void FlightRecorder::record_accelerator_version(
-    }
-  }
-  if (r.end_ != nullptr) {
-    bool completed = r.end_->query();
-    if (completed && !r.time_discovered_completed_) {
-      r.time_discovered_completed_ = c10::getTime();
-    }
-  }
+  update_state_impl(r);
 }
 
 std::vector<FlightRecorder::Entry> FlightRecorder::dump_entries() {
-  std::lock_guard<std::mutex> guard(mutex_);
-  std::vector<Entry> result;
-  result.reserve(entries_.size());
-  result.insert(
-      result.end(),
-      entries_.begin() + static_cast<std::ptrdiff_t>(next_),
-      entries_.end());
-  result.insert(
-      result.end(),
-      entries_.begin(),
-      entries_.begin() + static_cast<std::ptrdiff_t>(next_));
-  // query any remaining events
-  for (auto& r : result) {
-    update_state(r);
-    r.start_ = r.end_ = nullptr;
-  }
-  return result;
+  return dump_entries_impl<FlightRecorder::Entry>(entries_, next_, mutex_);
 }
 
 // Returns the entry with the given id, if it exists. Otherwise, returns
 // std::nullopt.
 std::optional<FlightRecorder::Entry> FlightRecorder::getEntry(
     std::optional<size_t> id) {
-  if (!enabled_ || !id) {
-    return std::nullopt;
-  }
-
-  std::unique_lock<std::mutex> guard(mutex_);
-  Entry entry = entries_.at(*id % max_entries_);
-  if (entry.id_ == *id) {
-    return entry;
-  } else {
-    return std::nullopt;
-  }
+  return getEntryImpl<FlightRecorder::Entry>(
+      entries_, enabled_, mutex_, max_entries_, id);
 }
 
 void FlightRecorder::retire_id(
     std::optional<size_t> id,
     bool compute_duration) {
-  if (!enabled_ || !id) {
-    return;
-  }
-
-  bool can_compute_duration = false;
-  Event* startEvent = nullptr;
-  Event* endEvent = nullptr;
-  std::optional<float> duration = std::nullopt;
-
-  std::unique_lock<std::mutex> guard(mutex_);
-
-  Entry* entry = &entries_.at(*id % max_entries_);
-  if (entry->id_ == *id) {
-    update_state(*entry);
-
-    if (compute_duration) {
-      can_compute_duration = entry->time_discovered_completed_.has_value() &&
-          entry->start_ && entry->end_;
-      startEvent = entry->start_;
-      endEvent = entry->end_;
-    }
-    entry->retired_ = true;
-    entry->start_ = entry->end_ = nullptr;
-  }
-
-  if (can_compute_duration) {
-    // Compute duration without without holding the lock, because
-    // cudaEventDuration() can hang, and we need to acquire the lock before we
-    // can dump(), which we never want to block.
-    guard.unlock();
-    duration = getDurationFromEvent(*startEvent, *endEvent);
-    guard.lock();
-
-    // Refresh the entry pointer, see if the entry has been overwritten
-    entry = &entries_.at(*id % max_entries_);
-    if (entry->id_ != *id) {
-      LOG(INFO) << "retire_id abandoned for id " << *id
-                << ", event was overwritten while waiting to compute duration.";
-      return;
-    }
-    if (duration.has_value()) {
-      entry->duration_ = duration;
-    }
-  }
+  retire_id_impl<FlightRecorder::Entry, Event>(
```

Comment:
Why do we need to use these inner functions? Seems like we don't actually reuse them anywhere? We can just make FlightRecorder::retire_id a templatized method right?

Is the concern that we would have to move this logic into the header file to make it templatized?